postProcessing/Folders/CosmicIntegration/PythonScripts/FastCosmicIntegration.py

```diff
@@ -406,23 +474,162 @@ def find_detection_rate(path, filename="COMPAS_Output.h5", dco_type="BBH", weigh
+            # The number of merging BBHs that need a weight
+            N_dco  = len(merger_rate[:,0])
+            
+            ####################
+            # binned_merger_rate will be the (observed) weights, binned by redshhift
+            binned_merger_rate    = np.zeros( (N_dco, len(redshift_bins)-1) )# create an empty list to fill
+            binned_detection_rate = np.zeros( (N_dco, len(redshift_bins)-1) )# create an empty list to fill
+
+            # loop over all redshift redshift_bins
+            for i in range(len(redshift_bins)-1):
+                binned_merger_rate[:,i]    = np.sum(merger_rate[:, digitized == i+1], axis = 1)
+                # only add detected rates for the 'detectable' redshiifts
+                if redshift_bins[i] < redshifts[n_redshifts_detection]:
+                    binned_detection_rate[:,i] = np.sum(detection_rate[:, digitized_det == i+1], axis = 1)
+            save_redshifts        = redshift_bins
+            save_merger_rate      = binned_merger_rate
+            save_detection_rate   = binned_detection_rate
+
+        else: 
+            #  To avoid huge filesizes, we don't really wan't All the data, 
+            # so we're going to save up to some redshift
+            z_index = np.digitize(maxz, redshifts) -1
+
+            # The detection_rate is a smaller array, make sure you don't go beyond the end
+            detection_index = z_index if z_index < n_redshifts_detection else n_redshifts_detection
+
+            print('You will only save data up to redshift ', maxz, ', i.e. index', redshifts[z_index])
+            save_redshifts        = redshifts
+            save_merger_rate      = merger_rate[:,:z_index]
+            save_detection_rate   = detection_rate[:,:detection_index]
+
+        print('save_redshifts', save_redshifts)
+
+        #################################################
+        # Write the rates as a seperate dataset
+        # re-arrange your list of rate parameters
+        DCO_to_rate_mask     = COMPAS.DCOmask #save this bool for easy conversion between BSE_Double_Compact_Objects, and CI weights
+        rate_data_list       = [DCO['SEED'][DCO_to_rate_mask], DCO_to_rate_mask , save_redshifts,  save_merger_rate, merger_rate[:,0], save_detection_rate]
+        rate_list_names      = ['SEED', 'DCOmask', 'redshifts',  'merger_rate','merger_rate_z0', 'detection_rate'+sensitivity]
+        for i, data in enumerate(rate_data_list):
+            print('Adding rate info of shape', np.shape(data))
+            # Check if dataset exists, if so, just delete it
+            if rate_list_names[i] in h_new[new_rate_group].keys():
+                del h_new[new_rate_group][rate_list_names[i]]
+            # write rates as a new data set
+            dataNew     = h_new[new_rate_group].create_dataset(rate_list_names[i], data=data)
+
+    #Always close your files again ;)
+    h_new.close()
+    print('Done with append_rates :) your new files are here: ', path + '/' + filename )
```

Comment:
Tom, is it always the case that the path will have a '/' at the end? If not, this could be a safe choice.

There are also python libraries that will handle correct joining of path names, which may be preferable.

Reply:
in my experience an extra '/' never hurts, while one too little could hurt! So I would argue to keep it

Reply:
Definitely a fair point! I think you could maybe get around this by doing `(path + '/' + filename).replace('//', '/')` (since I don't *think* paths should ever had a double slash).

But either way agreed that one too many is better than one too few!

Reply:
fair point! I've cleaned it up with this suggestion! :)